extensions/amp-iframe/1.0/component.js

```diff
@@ -26,23 +30,96 @@ export function Iframe({
   allowFullScreen,
   allowPaymentRequest,
   allowTransparency,
-  onLoadCallback = NOOP,
+  onLoad = NOOP,
   referrerPolicy,
+  requestResize,
   sandbox,
   src,
   srcdoc,
   ...rest
 }) {
+  const iframeRef = useRef();
+  const dataRef = useRef(null);
+  const isIntersectingRef = useRef(false);
+
+  const attemptResize = useCallback(() => {
+    const iframe = iframeRef.current;
+    if (!iframe) {
+      return;
+    }
+    const height = Number(dataRef.current.height);
+    const width = Number(dataRef.current.width);
+    if (!height && !width) {
+      return;
+    }
+    if (requestResize) {
+      // Currently `requestResize` is called twice:
+      // 1. when the message is received in viewport
+      // 2. when exiting viewport
+      // This could be optimized by reducing to one call by assessing when to call.
+      requestResize(height, width);
+      iframe.height = FULL_HEIGHT;
+      iframe.width = FULL_HEIGHT;
+    } else if (!isIntersectingRef.current) {
+      if (width) {
+        iframe.width = width;
+      }
+      if (height) {
+        iframe.height = height;
+      }
+    }
+  }, [requestResize]);
+
+  const handlePostMessage = useCallback(
+    (event) => {
+      // Currently we're only handling `embed-size` post messages
+      if (event.data?.type !== MessageType.EMBED_SIZE) {
+        return;
+      }
+      dataRef.current = event.data;
+      attemptResize();
+    },
+    [attemptResize]
+  );
+
+  useEffect(() => {
+    const iframe = iframeRef.current;
+    if (!iframe) {
+      return;
+    }
+    const win = iframe && toWin(iframe.ownerDocument.defaultView);
+    if (!win) {
+      return;
+    }
+    const io = new win.IntersectionObserver((entries) => {
```

Comment:
Feel free to punt on this idea nit: can we extract the InOb s.t. all `<amp-iframe>`s on the page will reuse the same instance? Or at least so that we don't create a new one each time the useEffect is called. Especially if different Bento components will start wanting to use an InOb for determining valid resizes it could be worth extracting into a custom hook (`const {boundingClientRect, inViewport, rootBounds} =  useInOb(ref)`)

Reply:
Good point. I will explore this in a separate PR.